Set "node.kubernetes.io/exclude-from-external-load-balancers" label at reboot and unset it at startup

ricky-rav opened the pull request:
Add label `node.kubernetes.io/exclude-from-external-load-balancers` on nodes that are going for a reboot and remove it at startup. This label is used by the cloud service controller in order to know which nodes to exclude from backend pools of cloud load balancers.

By adding this label at very start of the reboot process, instead of just waiting for the node to become NotReady, we give the cloud service controller more time to update backend pools on cloud load balancers.

This is part of the solution for BZ #2040715

squeed commented on pkg/daemon/update.go:
```diff
@@ -1988,10 +1992,91 @@ func (dn *Daemon) cancelSIGTERM() {
 // cleans up the agent's connections, and then sleeps for 7 days. if it wakes up
 // and manages to return, it returns a scary error message.
 func (dn *Daemon) reboot(rationale string) error {
+	dn.addExcludeFromLBLabel()
```
indeed, this should be in the `performDrain` function.

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,10 +1992,91 @@ func (dn *Daemon) cancelSIGTERM() {
+	err = dn.setExcludeFromLBLabel(false)
+	if err != nil {
+		dn.logSystem("Failed to remove label from node %s: %v", dn.name, err)
+		return nil
```
shouldn't this be returning error?

kikisdeliveryservice commented on pkg/daemon/daemon.go:
```diff
@@ -388,6 +388,9 @@ func (dn *Daemon) initializeNode() error {
 		return nil
 	}
+
+	dn.removeExcludeFromLBLabel()
```
presuming you're going to add err checking later on these funcs?

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,10 +1992,91 @@ func (dn *Daemon) cancelSIGTERM() {
+	// Get(ctx context.Context, name string, opts metav1.GetOptions) (*v1.Node, error)
+	// List(ctx context.Context, opts metav1.ListOptions) (*v1.NodeList, error)
+
+	return err
```
```suggestion
	return nil
```

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,10 +1992,91 @@ func (dn *Daemon) cancelSIGTERM() {
+	_, err = dn.kubeClient.CoreV1().Nodes().Patch(
+		context.TODO(), dn.name, types.MergePatchType, patchData, metav1.PatchOptions{})
+	if err != nil {
+		dn.logSystem("Error setting labels on node %s: %v", dn.name, err)
```
return err here and nil below

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,10 +1992,91 @@ func (dn *Daemon) cancelSIGTERM() {
+			dn.name, labelKey, labelValue, err)
+		return err
+	}
+	return err
```
```suggestion
	return nil
```

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,10 +1992,91 @@ func (dn *Daemon) cancelSIGTERM() {
+func (dn *Daemon) addExcludeFromLBLabel() error {
+	dn.setExcludeFromLBLabel(true)
+	// adding dummy file to mark that the label was added by MCD
+	dummyfile, e := os.Create(excludeExtLBPath)
```
```suggestion
	dummyfile, err := os.Create(excludeExtLBPath)
```

kikisdeliveryservice reviewed (commented):
A few comments on a first pass

deads2k commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1990,77 @@ func (dn *Daemon) cancelSIGTERM() {
+	if err != nil {
+		dn.logSystem("Failed to remove label from node %s: %v", dn.name, err)
+		return err
+	}
```
don't you need to delete the file before returning?

deads2k commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1990,77 @@ func (dn *Daemon) cancelSIGTERM() {
 	}
 }
 
+func (dn *Daemon) SetLabelsOnNode(labels map[string]interface{}) error {
```
why exposed?

kikisdeliveryservice replied:
+1 Also it's not setting _all_ labels on Node (as name indictes) it's only setting the LB labels on node.

deads2k commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1990,77 @@ func (dn *Daemon) cancelSIGTERM() {
+	return nil
+}
+
+func (dn *Daemon) addExcludeFromLBLabel() error {
```
missing test

deads2k commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1990,77 @@ func (dn *Daemon) cancelSIGTERM() {
+	return nil
+}
+
+func (dn *Daemon) setExcludeFromLBLabel(addLabel bool) error {
```
not sure about MCO standards, but I prefer to see LoadBalancer over LB

deads2k commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1990,77 @@ func (dn *Daemon) cancelSIGTERM() {
+	var labelValue interface{}
+	var err error
+	if addLabel {
+		labelValue = ""
```
Does the feature allow us to set a value to explain why we have set this label?

ricky-rav replied:
We could indeed set a value... The cloud service controller and in the az code only check for the presence of this label, without looking at the value.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L673
https://github.com/kubernetes/kubernetes/blob/b68063fce7abf62c69c257610d93ed3eb100956a/staging/src/k8s.io/legacy-cloud-providers/azure/azure.go#L850

kikisdeliveryservice commented on pkg/daemon/daemon.go:
```diff
@@ -154,6 +154,13 @@ const (
+	excludeFromLBLabel = "node.kubernetes.io/exclude-from-external-load-balancers"
+	// excludeExtLBPath is a dummy file marking that excludeFromLBLabel was added
+	// by MCD and not by the user.
+	excludeExtLBPath = "/etc/mco/added-exclude-from-external-load-balancers-label"
```
Ext->External

kikisdeliveryservice reviewed (commented):
+1 on deads2k comments, but also there are 3 funcs seemingly called the same/similar things, can we condense down to 1-2 for clarity and also add in text description of funcs?

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1990,77 @@ func (dn *Daemon) cancelSIGTERM() {
+func (dn *Daemon) setExcludeFromLBLabel(addLabel bool) error {
+	var labelValue interface{}
+	var err error
+	if addLabel {
```
Can you explain what's happening here with this bool? having var called addLabel true/false makes it sound like a label won't be added if false, but it's only setting a label value? T Which itself seems confusing here bc there doesn't seem to be any way for labelValue to have a value..

kikisdeliveryservice replied:
Tp just be clear, labelValue is declared above and is empty, if addLabel is true, it is..empty?

kikisdeliveryservice replied:
code changed since last push, trying to resolve conversation, but github keeps giving me an error

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1990,77 @@ func (dn *Daemon) cancelSIGTERM() {
+		dn.logSystem("File %s not present: nothing to do", excludeExtLBPath)
+		return nil
+	}
+	err = dn.setExcludeFromLBLabel(false)
```
Can you explain what this is doing? Why would we call the set function from remove? Can we do this without the bool usage?

deads2k commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1997,74 @@ func (dn *Daemon) cancelSIGTERM() {
 }
 
+// SetAnnotationsOnNode takes a map of key/value string pairs and patches them as labels on the node
+func (dn *Daemon) patchLabelsOnNode(labels map[string]interface{}) error {
```
Can you manually confirm that this patch will not overwrite *all* labels?

ricky-rav commented:
/bugzilla refresh

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1997,74 @@ func (dn *Daemon) cancelSIGTERM() {
 	}
 }
 
+// SetAnnotationsOnNode takes a map of key/value string pairs and patches them as labels on the node
```
this comment is outdated

ricky-rav commented:
> Can you manually confirm that this patch will not overwrite all labels?
This is still to be done. I'll paste the result of `oc label -oyaml ` in a comment in the code, as suggested by David.

> This is only used by your LB label so can we be clear on that in the naming?
Sure! I thought that keeping it generic was the best way of writing it, but I'm open to all suggestions :) 


So in the last commit I pushed:

- I moved `addOrRemoveExcludeFromLoadBalancerLabel` and `patchLabelsOnNode` to `daemon.go`, I thought it makes more sense there
- I added a fairly large unit test checking for `addOrRemoveExcludeFromLoadBalancerLabel`, where I also check that existing /different/ labels are untouched by the code.


Thanks!

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -1988,6 +1997,74 @@ func (dn *Daemon) cancelSIGTERM() {
 }
 
+// SetAnnotationsOnNode takes a map of key/value string pairs and patches them as labels on the node
+func (dn *Daemon) patchLabelsOnNode(labels map[string]interface{}) error {
```
on second thought it's fine to leave it as is.

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -59,6 +59,13 @@ const (
 	GPGNoRebootPath = "/etc/machine-config-daemon/no-reboot/containers-gpg.pub"
+
+	// excludeFromLoadBalancerLabel is used by the cloud service controller in order to know
+	// which nodes to exclude from backend pools of cloud load balancers.
```
Since you moved your funcs to daemon.go, these consts are orphaned. can we move them to live in daemon.go?

ricky-rav replied:
Yes, absolutely!

kikisdeliveryservice commented on pkg/daemon/daemon.go:
```diff
@@ -1629,3 +1634,71 @@ func forceFileExists() bool {
+		// removing the label by setting its value to nil
+		labelValue = nil
+	}
+	err = dn.patchLabelsOnNode(map[string]interface{}{excludeFromLoadBalancerLabel: labelValue})
```
Not sure @deads2k older comment was addressed here yet:
`for any other error on the stat, I would expect no action from this function.`

cheesesashimi commented on pkg/daemon/daemon.go:
```diff
@@ -1629,3 +1634,71 @@ func forceFileExists() bool {
+	labelValue, labelExists := dn.node.Labels[excludeFromLoadBalancerLabel]
+	if addLabel {
+		if labelExists {
+			// TODO remove log line below and make the if condition more compact
```
I don't mind this less-compact representation as it's clearer to reason about what is happening. Also, I would leave the log lines in place for now. If a more compact representation is desired, a `switch` block might be better than nested ifs.

ricky-rav replied:
Yes! Thanks for spotting that line, I actually forgot to remove it. My bad!!

cheesesashimi reviewed (changes requested):
Overall, this looks good. I do have a few suggestions though.

kikisdeliveryservice commented:
Just to summarize for anyone following: pending manual verification, passing ci and those 2 changes req by zack and i in our reviews, this is good to go.

Going to add my approval (since this is very close to being done) and a hold so that we can expedite merging pending the above.

/approve
/hold

ricky-rav commented:
I just manually verified that the the code:
1.  Adds the label unless it's present already (with whatever value)
2. Removes the label if and only if it carries the value set by MCD
3. Does not remove or alter existing labels on the node

I tested this on three worker nodes.

**Worker 1:**
No "exclude" label startup:
```
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: Standard_D4s_v3
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: centralus
    failure-domain.beta.kubernetes.io/zone: centralus-1
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus1-l2thg
    kubernetes.io/os: linux
    node-role.kubernetes.io/worker: ""
    node.kubernetes.io/instance-type: Standard_D4s_v3
    node.openshift.io/os_id: rhcos
    topology.disk.csi.azure.com/zone: centralus-1
    topology.kubernetes.io/region: centralus
    topology.kubernetes.io/zone: centralus-1
```

Corresponding logs:
`Feb 17 21:33:38 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus1-l2thg root[10211]: machine-config-daemon[2382]: label node.kubernetes.io/exclude-from-external-load-balancers not present on node ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus1-l2thg, nothing to remove`

When rebooting the label gets added:
```
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: Standard_D4s_v3
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: centralus
    failure-domain.beta.kubernetes.io/zone: centralus-1
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus1-l2thg
    kubernetes.io/os: linux
    node-role.kubernetes.io/worker: ""
    node.kubernetes.io/exclude-from-external-load-balancers: added-by-mco
    node.kubernetes.io/instance-type: Standard_D4s_v3
    node.openshift.io/os_id: rhcos
    topology.disk.csi.azure.com/zone: centralus-1
    topology.kubernetes.io/region: centralus
    topology.kubernetes.io/zone: centralus-1
```
Logs:
```
Feb 17 22:30:14 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus1-l2thg root[51862]: machine-config-daemon[2382]: Patching labels map[node.kubernetes.io/exclude-from-external-load-balancers:added-by-mco]
Feb 17 22:30:14 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus1-l2thg root[51863]: machine-config-daemon[2382]: Update prepared; beginning drain
```


After reboot the label gets removed:
```
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: Standard_D4s_v3
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: centralus
    failure-domain.beta.kubernetes.io/zone: centralus-1
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus1-l2thg
    kubernetes.io/os: linux
    node-role.kubernetes.io/worker: ""
    node.kubernetes.io/instance-type: Standard_D4s_v3
    node.openshift.io/os_id: rhcos
    topology.disk.csi.azure.com/zone: centralus-1
    topology.kubernetes.io/region: centralus
    topology.kubernetes.io/zone: centralus-1
```

Logs:
`Feb 17 22:32:56 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus1-l2thg root[4549]: machine-config-daemon[2719]: Patching labels map[node.kubernetes.io/exclude-from-external-load-balancers:<nil>]`

**Worker 2**
It is manually labeled with `"node.kubernetes.io/exclude-from-external-load-balancers": "mylabel"`.
Here are the relevant logs:
```
Feb 17 21:33:39 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus2-fqb8w root[11249]: machine-config-daemon[2385]: label node.kubernetes.io/exclude-from-external-load-balancers not present on node ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus2-fqb8w, nothing to remove

Feb 17 22:33:12 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus2-fqb8w root[69341]: machine-config-daemon[2385]: node ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus2-fqb8w already has label node.kubernetes.io/exclude-from-external-load-balancers=mylabel (addLabel=true), nothing to add
Feb 17 22:33:12 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus2-fqb8w root[69342]: machine-config-daemon[2385]: Update prepared; beginning drain

Feb 17 22:35:48 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus2-fqb8w root[4550]: machine-config-daemon[2881]: label node.kubernetes.io/exclude-from-external-load-balancers=mylabel was not added by MCO, not removing it
```
**Worker 3**
It is manually labeled with `"node.kubernetes.io/exclude-from-external-load-balancers": ""`.
Here are the relevant logs:

```
Feb 17 21:33:13 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus3-6dj98 root[10537]: machine-config-daemon[2368]: label node.kubernetes.io/exclude-from-external-load-balancers not present on node ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus3-6dj98, nothing to remove
Feb 17 22:36:04 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus3-6dj98 root[73477]: machine-config-daemon[2368]: node ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus3-6dj98 already has label node.kubernetes.io/exclude-from-external-load-balancers= (addLabel=true), nothing to add
Feb 17 22:39:16 ci-ln-0h2ssy2-1d09d-vpvjc-worker-centralus3-6dj98 root[5382]: machine-config-daemon[2867]: label node.kubernetes.io/exclude-from-external-load-balancers= was not added by MCO, not removing it

```

ricky-rav commented on pkg/daemon/daemon.go:
```diff
@@ -1629,3 +1634,71 @@ func forceFileExists() bool {
+		Metadata map[string]interface{} `json:"metadata"`
+	}{
+		Metadata: map[string]interface{}{
+			"labels": labels,
```
Thanks for your feedback! Can you please provide an example of how you see it? (I suppose it's straight forward, but at this hour I don't see it... :) )

The reason I use `labels map[string]interface{})` as the input type and later in the struct is that I need to pass it `nil` when I want to remove the label. Would I be able to do it with `Labels map[string]string` from `ObjectMeta`?

deads2k replied:
I attempted to do this here: https://github.com/openshift/machine-config-operator/pull/2954 , but the semantics of nil in the merge patch don't carry through with the switch to ObjectMeta.

deads2k replied:
> If not, labels is missing the json:"labels" tag and this could cause problems.

Because this is a map, the literal "labels" is used in the json.

deads2k replied:
it's not obvious, but there's a "normal" way to do this in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/label/label.go#L338 .  I would swap to it, but without an e2e test, the verification is harder for me.

cheesesashimi replied:
> Because this is a map, the literal "labels" is used in the json.

That's a great point. Thanks for the correction!

> I attempted to do this here: https://github.com/openshift/machine-config-operator/pull/2954 , but the semantics of nil in the merge patch don't carry through with the switch to ObjectMeta.

Between the nil semantics and your comment about the `labels` literal, I'll consider this resolved as-is.

ricky-rav commented:
/retest-required

cheesesashimi commented:
It looks like something in this PR is causing the `e2e-agnostic-upgrade` to consistently fail. To describe the failure mode in detail:

1. One of the control plane nodes goes down for an update.
2. The MachineConfig files get written to the node.
3. Retrieving the latest OS update continuously fails, causing the node to eventually become degraded and time out the test job.

I wrote a rudimentary e2e test that one could use to validate that this has the intended effects: https://github.com/openshift/machine-config-operator/pull/2958. An improved version of this e2e test should be included in this PR.

/hold
/lgtm cancel
/remove-approve

deads2k commented:
> Retrieving the latest OS update continuously fails, causing the node to eventually become degraded and time out the test job

Do you have an explanation for this behavior?  It seems strange that labelling a node would fail in this manner.  Is a process crashlooping somewhere?

kikisdeliveryservice commented on pkg/daemon/daemon.go:
```diff
@@ -1629,3 +1641,104 @@ func forceFileExists() bool {
+	var err error
+	var labelValue interface{}
+	// Add the label only if not present; remove the label only if it has our custom value
+	labelValue, labelExists := dn.node.Labels[excludeFromLoadBalancerLabel]
```
yeppp see:
https://github.com/openshift/machine-config-operator/blob/b68d8e1583e1b117e4fa968f0482e718d8d42f03/pkg/daemon/update.go#L356

you need to add a if dn.node{} around this conditional so it can pass in the firstboot where there is no node object.

kikisdeliveryservice commented on pkg/daemon/daemon.go:
```diff
@@ -1629,3 +1641,104 @@ func forceFileExists() bool {
+	var err error
+	var labelValue interface{}
+	// Add the label only if not present; remove the label only if it has our custom value
+	labelValue, labelExists := dn.node.Labels[excludeFromLoadBalancerLabel]
```
```suggestion
if dn.node != nil { 
	labelValue, labelExists := dn.node.Labels[excludeFromLoadBalancerLabel]
```

kikisdeliveryservice commented:
Latest change fixed yesterday's problem. Also noting that vsphere-upgrade did in fact pass (tho not the reqd agnostic upgrade), so this is probably ok with some retesting.

/retest-required

kikisdeliveryservice commented:
Just to followup, is this PR still needed?

squeed commented:
@kikisdeliveryservice @ricky-rav This PR would certainly be welcome; I think it would be a significant improvement in reliability.
